build-tools/packages/build-infrastructure/src/utils.ts

```diff
@@ -0,0 +1,83 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "node:path";
+import execa from "execa";
+
+import { NotInGitRepository } from "./errors.js";
+
+/**
+ * Returns the absolute path to the nearest Git repository root found starting at `cwd`.
+ *
+ * @param cwd - The working directory to use to start searching for Git repositories. Defaults to `process.cwd()` if not
+ * provided.
+ *
+ * @throws A `NotInGitRepository` error if no git repo is found.
+ *
+ * @privateRemarks
+ * This function is helpful because it is synchronous. The SimpleGit wrapper is async-only.
+ */
+export function findGitRootSync(cwd = process.cwd()): string {
+	try {
+		const result = execa.sync("git", ["rev-parse", "--show-toplevel"], {
+			cwd,
+			encoding: "utf8",
+			// Ignore stdin but pipe (capture) stdout and stderr since git will write to both.
+			stdio: ["ignore", "pipe", "pipe"],
+		});
+
+		// If anything was written to stderr, then it's not a git repo.
+		if (result.stderr) {
+			throw new NotInGitRepository(cwd);
+		}
+
+		return result.stdout.trim();
+	} catch (error) {
+		throw new NotInGitRepository(cwd);
```

Comment:
I'd remove the try-catch. A `NotInGitRepository()` will already propagate as such, and this is potentially hiding other kinds of errors. Better to let them propagate as whatever they are, IMO.

Reply:
The try/catch is needed because the execa call can throw, and there's no easy way to differentiate all the reasons it
could throw. Example of the error thrown (it's just an instance of Error):

Command failed with exit code 128: git rev-parse --show-toplevel
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).

Reply:
Added comments explaining.

Reply:
Then how about checking for "not a git repository" in the error message, throw `NotInGitRepository` in that case, and just rethrow otherwise? If git was not installed, for example, I'd rather see an error that makes that clear instead of thinking we're definitely not in a git repo.

Reply:
Done.